### DISCUSSION TITLE
DEV MODE is coming!

### DISCUSSION THREAD

**Discussion by Westbrook** [adobe/spectrum-web-components#2308] [Announcements]
If you've been following PRs to our repo then you might already know, but we've been hard at work to bring a DEV MODE to Spectrum Web Components. Both internally and externally we've heard of numerous opportunities for our elements to better support consuming developers in achieving the success they expect across a number of areas like accessibility, deprecations, element usage, internationalization, performance, and more, and we are getting close to having something available for teams to try out. This is also a big piece of functionality that we've been blocking our 1.0 release on, so it's exciting that it's getting so close.

## How does it work?
Spectrum Web Components DEV MODE will be accessed by leveraging the `development` [export condition](https://nodejs.org/api/packages.html#conditional-exports) when resolving the bare import specifiers that you are already using to access our elements. This means that the messages available in this more will be opt-in and completely excluded from your production builds, with no work on your part, when leveraging the "production" configuration of your build tools. When in DEV MODE, usage of elements that fall outside of expected parameters will publish a warning into the browser console to outline things that you might want to do in order to fully leverage the benefits of the Spectrum Web Components that you use to build your application when they are delivered to production.

## What does it do?
By default, nothing. Once you've added the `development` export condition to your environment, however, it will tell you that it is on:

<img width="439" alt="image" src="https://user-images.githubusercontent.com/1156657/169542533-ea3f6339-5699-4e9c-94fa-9e69667dedd3.png">

For elements, like `<sp-theme>`, that have complex usage expectations, you'll get information that will support you in leveraging them correctly:

<img width="439" alt="image" src="https://user-images.githubusercontent.com/1156657/169541875-b4a2d659-eeda-49c4-8abe-e5b3f6be2635.png">

When elements expect a specific combination of attributes or content to ensure that they are delivered accessibly, they'll let you know when those requirements do not seem to be met:

<img width="502" alt="image" src="https://user-images.githubusercontent.com/1156657/166256830-149535b1-cfba-4574-909b-4847d2dbd3ea.png">

Elements like `<sp-picker>` that have deprecated light DOM APIs will be able to warn you, only at development time, that you're doing so. As we continue to expand our support for internationalized and localized content, there are also places where we could point out that work in this area might be required on your part to ensure the delivery of your application across various locales. We're just starting to understand all the places where these sorts of messages may be useful to you, our consumers. 

## How can I help?
If you're interested in leveraging the DEV MODE in your application, we'd love to hear from you in the comments below.

**Have ideas on things that you wish had been easier or clearer when you were getting started?**
Share them below as they could be useful to the next engineer/team that takes up Spectrum Web Components in their project as we continue to see explosive growth in consumption both within Adobe and across the browser web ecosystem. 

**Found parts of our documentation that are confusing or lacking?**
You should definitely share, as we might need to make a change. But, if it's really just _tricky_ there might be a good way to support you and anyone else who's leveraging that pattern right where they're doing their work!

**Working on complex consumption of some element or elements?**
Share your usage below! The more we understand the usage across our various consumers the more powerful DEV MODE can be for everyone.

**Has your QA team caught issues in your consumption?**
Let us know, and hopefully, we can reduce dev cycle time when you add a similar feature in the future by calling out usage before you move your tasks to "in review".

## What's next?
Getting DEV MODE exports into the library is the main blocker we've identified in this work as we move toward a 1.0.0 release. However, once we get that into the library there will be so many opportunities to continue to iterate on the offering to better support developers like you. We're already starting to think about what that might look like:

- "full" coverage of issues across the various elements and patterns that we ship
- expand warnings to link out to the documentation site to complete the circle of information and expand on the whys and whats in our messaging
- a DEV MODE config to customize things like the types of messages you receive, the elements you receive them from, message types to ignore as desired, etc.
- what would you like to see?

If there are other types of development time messaging, or ideas around receiving them, that you'd like to see included in future work in this area, hop in the comments below. We look forward to hearing what you have got in mind and are excited to be able to bring this powerful new Spectrum Web Components feature to you and your applications, soon!

**Comment by Westbrook** [comment]
An initial pre-release for this feature has landed.

We’re starting with a small handful of messages, and will be expanding into useful areas with your help throughout the rest of the year. Starting with this release you’ll be able to get additional information about:

- The fact that you’re in Dev Mode. When you get this message, and no others, you can rest assured that we’ve not written helpers to catch whatever dirty little tricks you’ve been hiding in your code, yet. :wink:
- `<sp-badge>`
  - When you’re not leveraging the required variant attribute with a valid value. :name_badge:
- `<sp-picker>`
  - We’ve moved deprecation information about using <sp-menu> elements in conjunction with this element so that it will only ever appear at dev time.
- `<sp-progress-bar>` and `<sp-progress-circle>`
  - When you’re not leveraging the right attributes to ensure these elements are being delivered accessibly, dev mode will tell you! :guide_dog:
- `<sp-slider-handle>`
  - You will now be given additional information when the Slider Handle elements’ min and max values are not configured correctly.
- `<sp-theme>`
  - There are a number of attributes and fragment files that are needed to ensure that it delivers your theme correctly, and <sp-theme> will now let you know when your application is not prepared to deliver these correctly.


<details>

<summary>Find these changes in the releases below:</summary>

Successfully published:
 - @spectrum-web-components/accordion@0.6.15-devmode.0+07474d44f
 - @spectrum-web-components/action-bar@0.4.17-devmode.0+07474d44f
 - @spectrum-web-components/action-button@0.9.1-devmode.0+07474d44f
 - @spectrum-web-components/action-group@0.10.1-devmode.0+07474d44f
 - @spectrum-web-components/action-menu@0.14.1-devmode.0+07474d44f
 - @spectrum-web-components/asset@0.6.11-devmode.0+07474d44f
 - @spectrum-web-components/avatar@0.9.12-devmode.0+07474d44f
 - @spectrum-web-components/badge@0.1.4-devmode.0+07474d44f
 - @spectrum-web-components/banner@0.8.6-devmode.24+07474d44f
 - @spectrum-web-components/base@0.5.9-devmode.24+07474d44f
 - @spectrum-web-components/bundle@0.26.2-devmode.0+07474d44f
 - @spectrum-web-components/button-group@0.8.12-devmode.0+07474d44f
 - @spectrum-web-components/button@0.18.1-devmode.0+07474d44f
 - @spectrum-web-components/card@0.10.14-devmode.0+07474d44f
 - @spectrum-web-components/checkbox@0.13.6-devmode.0+07474d44f
 - @spectrum-web-components/clear-button@0.1.8-devmode.0+07474d44f
 - @spectrum-web-components/close-button@0.1.8-devmode.0+07474d44f
 - @spectrum-web-components/coachmark@0.9.8-devmode.0+07474d44f
 - @spectrum-web-components/color-area@0.4.11-devmode.0+07474d44f
 - @spectrum-web-components/color-handle@0.3.11-devmode.0+07474d44f
 - @spectrum-web-components/color-loupe@0.3.11-devmode.0+07474d44f
 - @spectrum-web-components/color-slider@0.3.13-devmode.0+07474d44f
 - @spectrum-web-components/color-wheel@0.3.13-devmode.0+07474d44f
 - @spectrum-web-components/dialog@0.10.9-devmode.0+07474d44f
 - @spectrum-web-components/divider@0.4.12-devmode.0+07474d44f
 - @spectrum-web-components/dropzone@0.9.11-devmode.0+07474d44f
 - @spectrum-web-components/field-group@0.5.12-devmode.0+07474d44f
 - @spectrum-web-components/field-label@0.7.14-devmode.0+07474d44f
 - @spectrum-web-components/help-text@0.1.11-devmode.0+07474d44f
 - @spectrum-web-components/icon@0.11.12-devmode.0+07474d44f
 - @spectrum-web-components/icons-ui@0.8.12-devmode.0+07474d44f
 - @spectrum-web-components/icons-workflow@0.8.12-devmode.0+07474d44f
 - @spectrum-web-components/icons@0.8.9-devmode.24+07474d44f
 - @spectrum-web-components/iconset@0.6.9-devmode.24+07474d44f
 - @spectrum-web-components/illustrated-message@0.8.13-devmode.0+07474d44f
 - @spectrum-web-components/link@0.12.8-devmode.0+07474d44f
 - @spectrum-web-components/menu@0.14.5-devmode.0+07474d44f
 - @spectrum-web-components/meter@0.6.14-devmode.0+07474d44f
 - @spectrum-web-components/modal@0.6.10-devmode.0+07474d44f
 - @spectrum-web-components/number-field@0.4.1-devmode.0+07474d44f
 - @spectrum-web-components/overlay@0.16.6-devmode.0+07474d44f
 - @spectrum-web-components/picker@0.11.6-devmode.0+07474d44f
 - @spectrum-web-components/popover@0.11.17-devmode.0+07474d44f
 - @spectrum-web-components/progress-bar@0.5.14-devmode.0+07474d44f
 - @spectrum-web-components/progress-circle@0.4.12-devmode.0+07474d44f
 - @spectrum-web-components/quick-actions@0.6.12-devmode.0+07474d44f
 - @spectrum-web-components/radio@0.9.15-devmode.0+07474d44f
 - @spectrum-web-components/search@0.11.6-devmode.0+07474d44f
 - @spectrum-web-components/shared@0.14.5-devmode.0+07474d44f
 - @spectrum-web-components/sidenav@0.12.15-devmode.0+07474d44f
 - @spectrum-web-components/slider@0.12.15-devmode.0+07474d44f
 - @spectrum-web-components/split-button@0.8.10-devmode.0+07474d44f
 - @spectrum-web-components/split-view@0.4.11-devmode.0+07474d44f
 - @spectrum-web-components/status-light@0.10.12-devmode.0+07474d44f
 - @spectrum-web-components/styles@0.15.1-devmode.0+07474d44f
 - @spectrum-web-components/swatch@0.1.2-devmode.0+07474d44f
 - @spectrum-web-components/switch@0.9.14-devmode.0+07474d44f
 - @spectrum-web-components/tabs@0.8.15-devmode.0+07474d44f
 - @spectrum-web-components/tags@0.8.15-devmode.0+07474d44f
 - @spectrum-web-components/textfield@0.11.11-devmode.0+07474d44f
 - @spectrum-web-components/theme@0.12.1-devmode.0+07474d44f
 - @spectrum-web-components/thumbnail@0.5.7-devmode.0+07474d44f
 - @spectrum-web-components/toast@0.10.14-devmode.0+07474d44f
 - @spectrum-web-components/tooltip@0.10.17-devmode.0+07474d44f
 - @spectrum-web-components/top-nav@0.5.15-devmode.0+07474d44f
 - @spectrum-web-components/tray@0.3.12-devmode.0+07474d44f
 - @spectrum-web-components/underlay@0.8.11-devmode.0+07474d44f
 - @spectrum-web-components/story-decorator@0.6.6-devmode.0+07474d44f
 - @spectrum-web-components/vrt-compare@0.1.17-devmode.0+07474d44f
 - @spectrum-web-components/grid@0.0.9-devmode.24+07474d44f
 - @spectrum-web-components/reactive-controllers@0.2.5-devmode.79+07474d44f
</details>

**Comment by Westbrook** [comment]
Latest message shape:

<img width="662" alt="image" src="https://user-images.githubusercontent.com/1156657/180089384-55ca7765-b8c1-49d3-87da-72ec8ac93cc5.png">

```
Default message:

- some list
- of pertinate
- info

Location of element in Elements Tab...

Link for more info.
Warning data for use in ignore settings.
```

What do you think?

**Comment by Westbrook** [comment]
This is shipping as of today!!! To start you'll see messages on the following packages:
- Bage, re: `variant` attribute usage
- Base, re: dev mode
- Picker, re: `sp-menu` deprecation
- Progress Bar/Circle, re: accessible consumption of their API
- Slider, re: min/max usage in slider handles
- Theme, re: attribute and fragment presence/consumption
Let us know what you think. Let us know what else you'd like to see. Look forward to updates in the coming months to support even more usage.

Enjoy! 🥳

**Comment by michaelwarren1106** [comment]
From one Design System dev to another, I have a question about the Dev Mode warning code. Is it stripped in the production file or still there just not run because of the global var check? If the warning code is stripped, how?! :) Im curious about implementing my own dev mode because helpful errors are great at dev time. Thanks!

**Reply by Westbrook** [reply]
The idea is to make the Dev Mode code "dead" so that it'll get minified away. We do this by leveraging esbuild to actually make two builds out of every TS file, one with `if (true) { / * dev mode stuff */ }` and one with `if (false) { /* dev mode stuff */ }` by way of their `define` option.
